source/language/openpulse.rst

```diff
@@ -0,0 +1,609 @@
+   // sigma is the standard deviation of waveform
+   // beta is the Y correction amplitude, see the DRAG paper
+   kernel drag(complex[size] amp, length l, length sigma, float[size] beta) -> waveform;
+
+   // amp is waveform amplitude
+   // l is the overall length of the waveform
+   kernel constant(complex[size] amp, length l) -> waveform;
+
+   // amp is waveform amplitude
+   // l is the overall length of the waveform
+   // frequency is the frequency of the waveform
+   // phase is the phase of the waveform
+   kernel sine(complex[size] amp, length  l, float[size] frequency, angle[size] phase) -> waveform;
+
+We can manipulate the ``waveform`` types using the following signal processing functions to produce
+new waveforms (this list may be updated as more functionality is required).
+
+.. code-block:: javascript
+
+    // Multiply two input waveforms entry by entry to produce a new waveform
+    // :math:`wf(t_i) = wf_1(t_i) \times wf_2(t_i)`
+    kernel mix(waveform wf1, waveform wf2) -> waveform;
+
+    // Sum two input waveforms entry by entry to produce a new waveform
+    // :math:`wf(t_i) = wf_1(t_i) + wf_2(t_i)`
+    kernel sum(waveform wf1, waveform wf2) -> waveform;
+
+    // Add a relative phase to a waveform (ie multiply by :math:`e^{\imag \theta}`
+    kernel phase_shift(waveform wf, angle ang) -> waveform;
+
+Play instruction
+----------------
+
+Waveforms are scheduled using the ``play`` instruction. These instructions may
+only appear inside a ``defcal`` block and have three required parameters:
+
+- the channel on which to play the pulse
+- a value of type ``waveform`` representing the waveform envelope
+- the frame to use for the pulse
+
+.. code-block:: javascript
+
+  play(channel chan, waveform wfm, frame fr)
+
+For example,
+
+.. code-block:: javascript
+
+  defcal play_my_pulses {
+   // Play a 3 sample pulse on the tx0 channel
+   play(tx0, [1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j], driveframe);
+
+   // Play a gaussian pulse on the tx1 channel
+   frame f1 = newframe(q1_freq, 0.0);
+   play(tx1, gaussian(...), f1);
+  }
+
+Capture Instruction
+-------------------
+
+Acquisition is scheduled by a ``capture`` instruction. This is a special
+``kernel`` function which is specified by a hardware vendor. The measurement
+process is difficult to describe generically due to the wide variety of
+hardware and measurement methods. Like the play instruction, these instructions
+may only appear inside a ``defcal`` block!
+
+The only required parameters are the ``channel`` and the ``frame``.
+
+The following are possible parameters that might be included:
+
+- A "duration" of type ``length``, if it cannot be inferred from other parameters.
+- A "filter" of type ``waveform``, which is dot product-ed with the measured IQ to distill the
+  result into a single IQ value
+
+Again it is up to the hardware vendor to determine the parameters and write a
+kernel definition at the top-level, such as:
```

Comment:
Again, all these `kernel`s need to become `extern`s.

Reply:
Addressed in same commit as others.